scipy/stats/_ksstats.py

```diff
@@ -0,0 +1,607 @@
+    if x <= 0.0:
+        return _select_and_pin_prob(0.0, 1.0, cdf=cdf)
+    if x >= 1.0:
+        return _select_and_pin_prob(1.0, 0.0, cdf=cdf)
+
+    pisquared = np.pi**2
+    pifour = np.pi**4
+    pisix = np.pi**6
+
+    z = np.sqrt(n) * x
+    zsquared = z**2
+    zfour = z**4
+    zsix = z**6
+
+    qlog = -pisquared / 8 / zsquared
+    if qlog < _MIN_LOG:  # z ~ 0.041743441416853426
+        return _select_and_pin_prob(0.0, 1.0, cdf=cdf)
+
+    q = np.exp(qlog)
+
+    # Coefficients of terms in the sums for K1, K2 and K3
+    k1a = -zsquared
+    k1b = pisquared / 4
+
+    k2a = 6 * zsix + 2 * zfour
+    k2b = (2 * zfour - 5 * zsquared) * pisquared / 4
+    k2c = pifour * (1 - 2 * zsquared) / 16
+
+    k3d = pisix * (5 - 30 * zsquared) / 64
+    k3c = pifour * (-60 * zsquared + 212 * zfour) / 16
+    k3b = pisquared * (135 * zfour - 96 * zsix) / 4
+    k3a = -30 * zsix - 90 * z**8
+
+    K0to3 = np.zeros(4)
+    # Use a Horner scheme to evaluate sum c_i q^(i^2)
+    # Reduces to a sum over odd integers.
+    maxk = int(np.ceil(16 * z / np.pi))
+    for k in range(maxk, 0, -1):
+        m = 2 * k - 1
+        msquared = m ** 2
+        mfour = m ** 4
+        msix = m ** 6
+        qpower = np.power(q, 8 * k)
+        coeffs = np.array([1.0,
+                           k1a + k1b * msquared,
+                           k2a + k2b*msquared + k2c * mfour,
+                           k3a + k3b*msquared + k3c * mfour + k3d * msix])
+        K0to3 *= qpower
+        K0to3 += coeffs
+    K0to3 *= q
+    K0to3 *= _SQRT2PI
+    # z**10 > 0 as z > 0.04
+    K0to3 /= np.array([z, 6 * zfour, 72 * z**7, 6480 * z**10])
+
+    # Now do the other sum over the other terms, all integers k
+    q = np.exp(-pisquared / 2 / zsquared)
+    # Coefficients of terms in the sums for K3
+    k2b = np.pi**2
+    k3b = 3 * pisquared * zsquared
+    k3c = -pifour
+    K0to3Extra = np.zeros(2)
+
+    for k in range(maxk, 0, -1):
+        ksquared = k**2
+        kfour = k**4
+        qpower = np.power(q, 2 * k + 1)
+        coeffs = np.array([k2b * ksquared, k3b * ksquared + k3c * kfour])
+        K0to3Extra *= qpower
+        K0to3Extra += coeffs
+    K0to3Extra *= q
+
+    zthree = z**3
+    K0to3Extra *= _SQRT2PI
+    K0to3Extra /= np.array([-36 * zthree, 216 * zsix])
+
+    K0to3[2:4] += K0to3Extra
+    powers_of_n = np.power(n * 1.0, np.arange(0, len(K0to3)) / 2.0)
+    K0to3 /= powers_of_n
+
+    if not cdf:
+        K0to3 *= -1
+        K0to3[0] += 1
+
+    Ksum = sum(K0to3)
+    return Ksum
+
+
+def _kolmogn(n, x, cdf=True):
+    """Compute the CDF(SF) for the two-sided Kolmogorov-Smirnov statistic.
+
+    x must be of type float, n of type integer.
+
+    Simard & L'Ecuyer (2011) [7]."""
+    if int(n) != n or n <= 0:
+        return np.nan
+    if x >= 1.0:
+        return _select_and_pin_prob(1.0, 0.0, cdf=cdf)
+    if x <= 0.0:
+        return _select_and_pin_prob(0.0, 1.0, cdf=cdf)
+    t = n * x
+    if t <= 1.0:  # Ruben-Gambino
```

Comment:
Ruben-Gambino certainly is valid for n>140. I replaced the product for n>=140 with a calculation using logs.

Reply:
> Ruben-Gambino certainly is valid for n>140

Indeed, I just noted that Simard/L'Ecuyer use RG for n <= 140 only. I think it is a good idea to use the asymptotic expansion for n > 140. is this covered by one of the test cases?

Reply:
The detailed descriptions in Simard/L'Ecuyer don't always explicitly list usage of R-G. 
From Figures 2 and 3 it is not immediately obvious whether the R-G regions stop at n=140, or extend beyond. A very close inspection suggests that they do indeed extend, but the line is very close to the y-axis.

But in Section 2.1 they imply that they use R-G whenever they can:

> In the tails (for x close to 0 or 1), the exact values for arbitrary n are (Ruben and Gambino 1982):
> ... (Formulae)
> This also provides an exact formula for the complementary cdf Fn(x) = 1 − Fn(x) over the same range of values of (n, x). We will use these simple formulas whenever they apply.

I.e. Simard/L'Ecuyer use R-G whenever it can be applied, and this implementation does the same.

Reply:
OK, Section 2.1 is pretty clear whereas 4.ii and 4.iii does not mention R-G. Thinking about it again, using R-G makes sense since one can accuratly estimate the very small probabilities using the approximation you implemented. Sorry about the confusion.